[filescanner] parse https streams for m3u

whatdoineed2do opened the pull request:
Fixes https://github.com/ejurgensen/forked-daapd/issues/608

Recognise `https` streams in `m3u` playlist files - the backend can play `https` streams fine with no further changes once patch applied

ejurgensen commented:
Have you tested this with a https m3u or pls? So make a m3u and then in that insert a link to https://xxx/xxx.m3u. There will probably be some sort of failure, I just want to make sure it is sane.

whatdoineed2do commented:
Tested, yes for m3u file that contained this:

```
https://cdn-storage.br.de/iLCpbHJGNL9zu6i6NL97bmWH_-bg/_-ZS/_AFH5-k69U1S/82362b63-f62d-4656-affc-c7becba1eaaa_2.mp3
```
It plays a german stream.

ejurgensen commented:
That's an mp3, I'm asking for a test of a m3u link inside an m3u file, where the link to the m3u file is https.

whatdoineed2do commented:
I misunderstood your request - I had to test this on a local server but it works for the case you describe -- although digging around I dont see any `https://.../...m3u` playlists.

Yes it works.

* forked daapd m3u playlist
```
https://localhost/foo.m3u
http://bbcmedia.ic.llnwd.net/stream/bbcmedia_radio1_mf_p
http://bbcmedia.ic.llnwd.net/stream/bbcmedia_radio2_mf_p
http://bbcmedia.ic.llnwd.net/stream/bbcmedia_radio4fm_mf_p
http://bbcmedia.ic.llnwd.net/stream/bbcmedia_radio4extra_mf_p
#EXTINF:-1,RTHK - RTHK Radio 2
http://stm.rthk.hk/radio2
http://stm.rthk.hk/radio1
http://stm.rthk.hk/radio3
https://cdn-storage.br.de/iLCpbHJGNL9zu6i6NL97bmWH_-bg/_-ZS/_AFH5-k69U1S/82362b63-f62d-4656-affc-c7becba1eaaa_2.mp3
```
* content of `https://localhost/foo.m3u` and donwloadable via `curl`
```
http://bbcwssc.ic.llnwd.net/stream/bbcwssc_mp1_ws-eieuk
```
The local http server (lighttpd) has its own fake CA and signed cert (see https://gist.github.com/fntlnz/cf14feb5a46b2eda428e000157447309 and https://www.digicert.com/ssl-certificate-installation-lighttpd.htm).

The server log:
```
[  LOG]     scan: Bulk library scan completed in 1 sec
[  LOG]      lib: Library init scan completed in 1 sec (2 changes)
[  LOG]     scan: Modified playlist found, processing '/export/public/music/foo.m3u'
[ INFO]     http: Making request for https://localhost/foo.m3u
[  LOG]    cache: Beginning DAAP cache update
[  LOG]    cache: DAAP cache updated
[ WARN]     scan: ffmpeg/libav could not extract any metadata
[ WARN]     scan: ffmpeg/libav could not extract any metadata
[ WARN]     scan: ffmpeg/libav could not extract any metadata
[ WARN]     scan: ffmpeg/libav could not extract any metadata
[ INFO]       db: Waiting for database unlock
[ WARN]     scan: ffmpeg/libav could not extract any metadata
[ WARN]     scan: ffmpeg/libav could not extract any metadata
[ WARN]     scan: ffmpeg/libav could not extract any metadata
[ WARN]     scan: ffmpeg/libav could not extract any metadata
[ WARN]     scan: ffmpeg/libav could not extract any metadata
[  LOG]     scan: Done processing playlist, added/modified 9 items
[  LOG]    cache: Beginning DAAP cache update
[  LOG]    cache: DAAP cache updated
[ INFO]   player: Opening 'https://localhost/foo.m3u' (id=26, item-id=517)
[ INFO]     http: Making request for https://localhost/foo.m3u
[  LOG]   laudio: open '/dev/snd/pcmC0D0p' failed (-77)[ INFO]   player: Using selected ALSA device 'Computer'
[ INFO]     raop: Failed to retrieve artwork for file id 517; no artwork will be sent
[ INFO]   laudio: Taking action to compensate for ALSA latency of 762 samples
[ INFO]   laudio: Taking action to compensate for ALSA latency of 780 samples
```

ejurgensen commented:
Do you have the option of blocking http/port 80? I think it may actually be retrieving the file via http, despite the log message.

ejurgensen commented:
Or alternatively try with Google Drive or Dropbox. I imagine they don't allow http.

whatdoineed2do commented:
Instead of the dropbox/gdrive's links, I found this https://..m3u
`https://www.internet-radio.com/servers/tools/playlistgenerator/?u=http://us4.internet-radio.com:8266/listen.pls&t=.m3u`

Forked daapd does like but it looks like its the ffmpeg doesnt like it:
```
[ INFO]     scan: Added new playlist as id 28
[ WARN]     scan: Cannot open media file 'https://www.internet-radio.com/servers/tools/playlistgenerator/?u=http://us4.internet-radio.com:8266/listen.pls&t=.m3u': Invalid data found when processing input
[  LOG]     scan: Playlist URL 'https://www.internet-radio.com/servers/tools/playlistgenerator/?u=http://us4.internet-radio.com:8266/listen.pls&t=.m3u' is unavailable for probe/metadata, assuming MP3 encoding
[ INFO]     http: Making request for https://localhost/foo.m3u
[ WARN]     scan: ffmpeg/libav could not extract any metadata
[ WARN]     scan: ffmpeg/libav could not extract any metadata
[ WARN]     scan: ffmpeg/libav could not extract any metadata
[ WARN]     scan: ffmpeg/libav could not extract any metadata
[ INFO]       db: Waiting for database unlock
[ WARN]     scan: ffmpeg/libav could not extract any metadata
[ WARN]     scan: ffmpeg/libav could not extract any metadata
[ WARN]     scan: ffmpeg/libav could not extract any metadata
[ WARN]     scan: ffmpeg/libav could not extract any metadata
[ WARN]     scan: ffmpeg/libav could not extract any metadata
[  LOG]     scan: Done processing playlist, added/modified 10 items
[  LOG]       db: Could not create query, unknown type 4
[  LOG]     scan: Skipping /export/public/music/.foo.m3u.swp, lstat() failed: No such file or directory
[  LOG]     scan: Skipping /export/public/music/.foo.m3u.swp, read_attributes() failed
[  LOG]      web: No queue item at position '0'
[  LOG]       db: Could not create query, unknown type 4
[  LOG]    cache: Beginning DAAP cache update
[  LOG]    cache: DAAP cache updated
[ INFO]   player: Opening 'https://www.internet-radio.com/servers/tools/playlistgenerator/?u=http://us4.internet-radio.com:8266/listen.pls&t=.m3u' (id=29, item-id=689)
[ WARN]   player: *** Input type=1 kind=1
[ WARN]   ffmpeg: Will reconnect at 36 in 0 second(s), error=End of file.
[ WARN]   ffmpeg: Will reconnect at 36 in 0 second(s), error=End of file.
[ WARN]   ffmpeg: Will reconnect at 36 in 0 second(s), error=End of file.
[ WARN]   ffmpeg: Will reconnect at 36 in 0 second(s), error=End of file.
[ WARN]   ffmpeg: Will reconnect at 36 in 0 second(s), error=End of file.
[ WARN]   ffmpeg: Will reconnect at 36 in 0 second(s), error=End of file.
[ WARN]   ffmpeg: Will reconnect at 36 in 0 second(s), error=End of file.
[ WARN]   ffmpeg: Will reconnect at 36 in 0 second(s), error=End of file.
[ WARN]   ffmpeg: Will reconnect at 36 in 0 second(s), error=End of file.
[ WARN]   ffmpeg: Will reconnect at 36 in 0 second(s), error=End of file.
[ WARN]   ffmpeg: Will reconnect at 36 in 0 second(s), error=End of file.
[ WARN]   ffmpeg: Will reconnect at 36 in 0 second(s), error=End of file.
[ WARN]   ffmpeg: Will reconnect at 36 in 0 second(s), error=End of file.
[ WARN]   ffmpeg: Will reconnect at 36 in 0 second(s), error=End of file.
[ WARN]   ffmpeg: Will reconnect at 36 in 0 second(s), error=End of file.
[ WARN]   ffmpeg: Will reconnect at 36 in 0 second(s), error=End of file.
[ WARN]   ffmpeg: Will reconnect at 36 in 0 second(s), error=End of file.
[ WARN]   ffmpeg: Will reconnect at 36 in 0 second(s), error=End of file.
[ WARN]   ffmpeg: Will reconnect at 36 in 0 second(s), error=End of file.
[ WARN]   ffmpeg: Will reconnect at 36 in 0 second(s), error=End of file.
[ WARN]   ffmpeg: Will reconnect at 36 in 0 second(s), error=End of file.
[ WARN]   ffmpeg: Will reconnect at 36 in 0 second(s), error=End of file.
[ WARN]   ffmpeg: Will reconnect at 36 in 0 second(s), error=End of file.
[ WARN]   ffmpeg: Will reconnect at 36 in 0 second(s), error=End of file.
[ WARN]   ffmpeg: Will reconnect at 36 in 0 second(s), error=End of file.
[ WARN]   ffmpeg: Will reconnect at 36 in 0 second(s), error=End of file.
[ WARN]   ffmpeg: Will reconnect at 36 in 0 second(s), error=End of file.
[ WARN]   ffmpeg: Will reconnect at 36 in 0 second(s), error=End of file.
[ WARN]   ffmpeg: Will reconnect at 36 in 0 second(s), error=End of file.
[ WARN]   ffmpeg: Will reconnect at 36 in 0 second(s), error=End of file.
[  LOG]    xcode: Timeout while reading source (connection problem?)
[ WARN]   ffmpeg: Will reconnect at 36 in 1 second(s), error=End of file.
[  LOG]    xcode: Timeout while reading source (connection problem?)
[  LOG]    xcode: Timeout while reading source (connection problem?)
[  LOG]    xcode: Cannot open 'https://www.internet-radio.com/servers/tools/playlistgenerator/?u=http://us4.internet-radio.com:8266/listen.pls&t=.m3u': Immediate exit requested
[  LOG]   player: Failed to open 'https://www.internet-radio.com/servers/tools/playlistgenerator/?u=http://us4.internet-radio.com:8266/listen.pls&t=.m3u' (id=29, item-id=689)
```

running `ffmpeg` util on the https link itself gives the same issue:
```
$ ffmpeg -hide_banner -i "https://www.internet-radio.com/servers/tools/playlistgenerator/?u=http://us4.internet-radio.com:8266/listen.pls&t=.m3u" -f null -

...

Opening an input file: https://www.internet-radio.com/servers/tools/playlistgenerator/?u=http://us4.internet-radio.com:8266/listen.pls&t=.m3u.
[NULL @ 0x55becfa42e00] Opening 'https://www.internet-radio.com/servers/tools/playlistgenerator/?u=http://us4.internet-radio.com:8266/listen.pls&t=.m3u' for reading
[https @ 0x55becfa43880] Setting default whitelist 'http,https,tls,rtp,tcp,udp,crypto,httpproxy'
[https @ 0x55becfa43880] request: GET /servers/tools/playlistgenerator/?u=http://us4.internet-radio.com:8266/listen.pls&t=.m3u HTTP/1.1
User-Agent: Lavf/58.12.100
Accept: */*
Range: bytes=0-
Connection: close
Host: www.internet-radio.com
Icy-MetaData: 1


[https @ 0x55becfa43880] header='HTTP/1.1 200 OK'
[https @ 0x55becfa43880] http_code=200
[https @ 0x55becfa43880] header='Server: nginx/1.14.0'
[https @ 0x55becfa43880] header='Date: Fri, 26 Oct 2018 19:14:16 GMT'
[https @ 0x55becfa43880] header='Content-Type: audio/mpegurl'
[https @ 0x55becfa43880] header='Content-Length: 36'
[https @ 0x55becfa43880] header='Connection: close'
[https @ 0x55becfa43880] header='Content-Disposition: attachment; filename="internet-radio.com.playlist.m3u"'
[https @ 0x55becfa43880] header='X-Powered-By: Perl Dancer 1.3202'
[https @ 0x55becfa43880] header=''
[AVIOContext @ 0x55becfdaa3c0] Statistics: 36 bytes read, 0 seeks
https://www.internet-radio.com/servers/tools/playlistgenerator/?u=http://us4.internet-radio.com:8266/listen.pls&t=.m3u: Invalid data found when processing input
```
Using `http` is fine so I wonder if `ffmpeg` is not expecting `https` based m3u streams.
`ffmpeg -hide_banner -i "http://us4.internet-radio.com:8266/listen.pls&t=.m3u" -f null -`

ejurgensen commented:
Thanks for thorough testing. I think it is ok, so I have merged it now.
